src/Compilers/Core/Portable/Symbols/Attributes/UnmanagedCallersOnlyAttributeData.cs

```diff
@@ -0,0 +1,33 @@
+
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis.Symbols;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class UnmanagedCallersOnlyAttributeData
+    {
+        internal static readonly UnmanagedCallersOnlyAttributeData Uninitialized = new UnmanagedCallersOnlyAttributeData(callingConventionTypes: null, isValid: false);
+
+        public readonly ImmutableHashSet<INamedTypeSymbolInternal>? CallingConventionTypes;
+
+#pragma warning disable 8775 // Invariant is verified in the constructor
+        [MemberNotNullWhen(true, nameof(CallingConventionTypes))]
+        public bool IsValid { get; }
+#pragma warning restore 8775
+
+        public UnmanagedCallersOnlyAttributeData(ImmutableHashSet<INamedTypeSymbolInternal>? callingConventionTypes, bool isValid)
```

Comment:
>? [](start = 91, length = 1)

Is it necessary to allow `null`?

Reply:
The only place that's used is in `Uninitialized`. I could create a private constructor for that to use and remove the attribute from this one.

---
In reply to: [482243461](https://github.com/dotnet/roslyn/pull/47311#discussion_r482243461) [](ancestors = 482243461)

Reply:
> I could create a private constructor for that to use and remove the attribute from this one.

Could we pass `ImmutableHashSet<INamedTypeSymbolInteral>.Empty` for `Uninitialized` instead, and simplify uses of this property?

---
In reply to: [482277886](https://github.com/dotnet/roslyn/pull/47311#discussion_r482277886) [](ancestors = 482277886,482243461)

Reply:
We could. I wanted it to be truly invalid so that dereferences would cause unexpected exceptions, but I could just make it empty instead.

---
In reply to: [482289635](https://github.com/dotnet/roslyn/pull/47311#discussion_r482289635) [](ancestors = 482289635,482277886,482243461)